### PR TITLE
Rename refactoring

### PR DESCRIPTION
**Pull request by dhuebner**
Default rename refactoring implementation which works on a declaration and cross references.
It is based on the index branch hence supports cross file rename refactoring.

### CODE REVIEW

**Comment by spoenemann** [comment]
Depends on #172.

**Comment by dhuebner** [comment]
Adapted to the new master branch changes

File: packages/langium/src/lsp/rename-refactoring.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    renameElement(document: LangiumDocument, params: RenameParams): WorkspaceEdit | undefined {
+        const textChanges: { [uri: string]: TextEdit[] } = {};
+        this.referenceFinder.findReferences(document, params, true).forEach((location) => {
+            textChanges[location.uri] ?
+                textChanges[location.uri].push(TextEdit.replace(location.range, params.newName))
+                :
+                textChanges[location.uri] = [TextEdit.replace(location.range, params.newName)];
+        });
+        return { changes: textChanges };
+    }
```
Comment (msujew) [inline packages/langium/src/lsp/rename-refactoring.ts:43]: Nitpick: The formatting here looks a bit off
```suggestion
    }
    
    renameElement(document: LangiumDocument, params: RenameParams): WorkspaceEdit | undefined {
        const changes: Record<string, TextEdit[]> = {};
        this.referenceFinder.findReferences(document, params, true).forEach(location => {
            textChanges[location.uri]
                ? textChanges[location.uri].push(TextEdit.replace(location.range, params.newName))
                : textChanges[location.uri] = [TextEdit.replace(location.range, params.newName)];
        });
        return { changes };
    }
```

File: packages/langium/src/lsp/language-server.ts [hunk 2]
```diff
@@ -162,6 +165,26 @@ export function addDocumentHighlightsHandler(connection: Connection, services: L
     });
 }
 
+export function addRenameHandler(connection: Connection, services: LangiumServices): void {
+    const renameHandler = services.lsp.RenameHandler;
+    connection.onRenameRequest((params: RenameParams): WorkspaceEdit | undefined => {
+        const document = paramsDocument(params, services);
+        if (document) {
+            return renameHandler?.renameElement(document, params);
```
[the remark this answers is not in the document]
Comment (dhuebner) [reply packages/langium/src/lsp/language-server.ts:173]: Rename handler was optional in the first draft. That is not true now as we have a default implementation

[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/lsp/language-server.ts:173]: Good to know, would you mind removing both of the `?` operators then?

File: packages/langium/src/lsp/rename-refactoring.ts (hunk 1 above)
Comment (dhuebner) [reply packages/langium/src/lsp/rename-refactoring.ts]: Here is more than just formatting changed :-/

File: packages/langium/src/lsp/rename-refactoring.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                return toRange(leafNode, doc);
+            }
+        }
+        return undefined;
+    }
+
+    protected isNameNode(leafNode: CstNode | undefined): boolean | undefined {
+        return leafNode?.element && isNamed(leafNode.element) && leafNode === this.nameProvider.getNameNode(leafNode.element);
+    }
+}
```
[the remark this answers is not in the document]
Comment (dhuebner) [reply packages/langium/src/lsp/rename-refactoring.ts]: Done. Would it be complicated to create an eslint for that?

**Comment by dhuebner** [comment]
@msujew 
Thanks a lot for your review!
All changes are applied

**Review by msujew — APPROVED** [review]
Renaming works mostly well. We already noticed that the reference finder doesn't work particularly well with the new indexing, but that's a topic for another PR.

LGTM!

**Comment by dhuebner** [comment]
@msujew 
I would like to track the review changes in a separate commit.
Is keeping two commits also okay?

**Comment by msujew** [comment]
I don't see a reason why review changes belong in a separate commit. I get it for separate features/parts of features that are implemented in the same PR, but not for this.

**Comment by dhuebner** [comment]
I would like to track things like `Co-authored-by: Mark Sujew <mark.sujew@typefox.io>` and see the changes I did because of the review. Maybe it's a bit strange but it gives me a better impression what and why happened :)

**Pull request merged**